File: MealTracker/Service Layer/UserService.m
//
//  UserService.m
//  MealTracker
//
//  Created by William Hindenburg on 2/17/15.
//
//

#import <BaseClasses/ServiceClient.h>
#import <BaseClasses/User.h>

#import "UserService.h"
#import <BaseClasses/RestKit/RestKit.h>

@interface UserService()
@property (strong, nonatomic) NSString *objectId;
@end

@implementation UserService

- (void)updateUser:(User *)user withSuccessBlock:(void (^)(User *user))successBlock andError:(void (^)(NSError *error))errorBlock {
    self.objectId = user.objectId;
    ServiceClient *serviceClient = [ServiceClient new];
    [serviceClient putObject:user andService:self withSuccessBlock:^(RKMappingResult *result) {
        successBlock(user);
    } andError:^(NSError *error) {
        errorBlock(error);
    }];
}

- (NSString *)serviceURL {
    return [NSString stringWithFormat:@"/1/users/%@", self.objectId];
}

- (NSString *)rootKeyPath {
    return nil;
}

- (NSString *)rootRequestKeyPath {
    return nil;
}

- (RKObjectMapping *)serializedMappingProvider {
    RKObjectMapping *mapping = [RKObjectMapping mappingForClass:[User class]];
    [mapping addAttributeMappingsFromDictionary:@{@"objectId":@"objectId"}];
    [mapping addAttributeMappingsFromDictionary:@{@"sessionToken":@"sessionToken"}];
    [mapping addAttributeMappingsFromDictionary:@{@"username":@"username"}];
    [mapping addAttributeMappingsFromDictionary:@{@"pointsPerWeek":@"pointsPerWeek"}];
    return mapping;
}

@end

File: MealTracker/Service Layer/UserService.swift
//
//  UserService.swift
//  MealTracker
//
//  Created by William Hindenburg on 1/10/16.
//
//

import UIKit
import BaseClassesSwift

class UserService: NSObject, BaseClassesService {
    var objectId = ""
    
    func updateUser(_ user:User, successBlock:@escaping ((User) -> Void), errorBlock:@escaping ((NSError) -> Void)) {
        self.objectId = user.objectId
        BaseClassesServiceClient().putObject(user, andService: self, successBlock: { (user:User) -> Void in
            successBlock(user)
        }) { (error) -> Void in
            errorBlock(error)
        }
    }
    
    func updatePoints(_ points:Points, successBlock:@escaping ((Points) -> Void), errorBlock:@escaping ((NSError) -> Void)) {
        BaseClassesServiceClient().postObject(points, andService: self, successBlock: { (points:Points) -> Void in
            successBlock(points)
        }) { (error) -> Void in
            errorBlock(error)
        }
    }
    
    func getPoints(_ successBlock:@escaping ((Points) -> Void), errorBlock:@escaping ((NSError) -> Void)) {
        BaseClassesServiceClient().getObject(self, successBlock: { (points:Points) in
            successBlock(points)
        }) { (error) in
            errorBlock(error)
        }
    }
    
    var serviceURL:String {
        return "/weeklypoints"
    }
}
